backends/ebpf/ebpfParser.cpp

```diff
@@ -335,8 +326,79 @@ bool StateTranslationVisitor::preorder(const IR::MethodCallExpression* expressio
         }
+    } else if (mi->is<P4::ExternFunction>()) {
+        auto func = mi->to<P4::ExternFunction>();
+        if (func->method->name.name == "assert") {
+            BUG_CHECK(expression->arguments->size() == 1,
+                    "%1%: Expected one argument ", expression);
+            builder->emitIndent();
+            builder->append("if (");
+            visit(expression->arguments->at(0)->expression);  // condition
+            builder->append(" == false) ");
+            builder->blockStart();
+            builder->emitIndent();
+            builder->appendFormat("% s = % s;", state->parser->program->errorVar.c_str(),
+                P4::P4CoreLibrary::instance.assertError);
+            builder->newline();
+            auto srcInfo = expression->srcInfo;
+            auto sourceFragment = srcInfo.toBriefSourceFragment();
+            auto sourceFile = srcInfo.getSourceFile();
+            auto linePosition = srcInfo.toPositionString();
+            builder->emitIndent();
+            builder->appendFormat(
+                "fprintf(stderr, \"Assertion error: '% s' failed, file '% s', line % s\\n\");",
+                sourceFragment.c_str(), sourceFile.c_str(), linePosition.c_str());
+            builder->newline();
+            builder->emitIndent();
+            builder->appendFormat("goto %s;", IR::ParserState::reject.c_str());
+            builder->newline();
+            builder->blockEnd(true);
+            return false;
+        } else if (func->method->name.name == "assume") {
+            BUG_CHECK(expression->arguments->size() == 1,
+                    "%1%: Expected one argument ", expression);
+            builder->emitIndent();
+            builder->append("if (");
+            visit(expression->arguments->at(0)->expression);  // condition
+            builder->append(" == true) ");
+            builder->blockStart();
+            builder->emitIndent();
+            builder->appendFormat("% s = % s;", state->parser->program->errorVar.c_str(),
+                P4::P4CoreLibrary::instance.assumeError);
+            builder->newline();
+            auto srcInfo = expression->srcInfo;
+            auto sourceFragment = srcInfo.toBriefSourceFragment();
+            auto sourceFile = srcInfo.getSourceFile();
+            auto linePosition = srcInfo.toPositionString();
+            builder->emitIndent();
+            builder->appendFormat(
+                "fprintf(stderr, \"Assume error: '% s' failed, file '% s', line % s\\n\");",
+                sourceFragment.c_str(), sourceFile.c_str(), linePosition.c_str());
+            builder->newline();
+            builder->emitIndent();
+            builder->appendFormat("goto %s;", IR::ParserState::reject.c_str());
+            builder->newline();
+            builder->blockEnd(true);
+            return false;
+        }
+    } else if (mi->is<P4::BuiltInMethod>()) {
+    // Added this because compiler didn't recognize assert(hdr.ethernet.isValid()), for expample
```

Comment:
if this is a bug maybe you can file a separate PR and a test case for it.
When does the compiler not recognize isValid, only within assert? 
isValid should be handled by the expressionConverter.

Reply:
Compiler didn't recognize isValid in assert, but also it didn't recognize setValid and setInvalid method calls for that ebpf backend when they are called from a parser state.
Also, I notice that compiler doesn't recognize verify call in parser state for ebpf backend, but I didn't add code generation for that case. In that case following error is reported:
error: Unexpected method call in parser verify
 verify(false, error.NoError);
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
If I understood it well, it should be enabled to call  methods, that are mentioned above, in parser state.

Reply:
It's possible that verify is not supported yet in the back-end. 
But isValid should work; if you have a counter-example please file another issue.